Add `avoid-accidental-submissions` feature

darkred opened the pull request:
<!--

Thanks for contributing! 🍄 Do not ignore this template, plz.

1. Does this PR close/fix an existing issue? Write something like `Closes #10`

Help us test and visualize this PR:

2. What pages does this PR affect? Include some REAL URLs where you tested the code
3. If applicable, add demonstrative screenshots or gifs

Lastly:

4. Open the PR as draft and review it yourself first. Fix what you find and explain weird code, if necessary.

-->

Closes #4248 

## Test URLs

- New Issue: https://github.com/sindresorhus/refined-github/issues/new
- New PR: https://github.com/sindresorhus/refined-github/compare/main...darkred:remap-enter-key-on-issue-pr-commit-title-fields?expand=1
- New commit:
  - Create new file: https://github.com/sindresorhus/refined-github/new/main
  - Modify an existing file: https://github.com/sindresorhus/refined-github/edit/main/readme.md

## Screenshot
![](https://user-images.githubusercontent.com/723651/115148453-5818df00-a068-11eb-92d4-51ac6e937b8b.gif)

fregante commented on readme.md:
```diff
@@ -176,6 +176,8 @@ Thanks for contributing! 🦋🙌
 - [](# "table-input") [Adds a button in the text editor to quickly insert a simplified HTML table.](https://user-images.githubusercontent.com/46634000/94559114-09892c00-0261-11eb-8fb0-c5a85ea76b6f.gif)
 - [](# "unfinished-comments") [Notifies the user of unfinished comments in hidden tabs.](https://user-images.githubusercontent.com/1402241/97792086-423d5d80-1b9f-11eb-9a3a-daf716d10b0e.gif)
 - [](# "wait-for-attachments") [Wait for the attachments to finish uploading before allowing to post a comment.](https://user-images.githubusercontent.com/46634000/104294547-9b8b0c80-54bf-11eb-93e5-65ae158353b3.gif)
+- [](# "avoid-accidental-submissions") [Re-maps the enter key to act as a tab in issue/PR/commit title fields, so it will merely move the cursor to the textarea
+.](https://user-images.githubusercontent.com/723651/115148453-5818df00-a068-11eb-92d4-51ac6e937b8b.gif)
```
It needs to be on one line and must explain why it does that (----> to avoid accidental submissions)

darkred replied:
Thank you. I removed the newline and added the explanation.

fregante commented:
Thank you for the PR and for following the template! I fixed some mistakes but I'm not sure it passes lint/types.

yakov116 commented:
Can't we `onConversationHeader` something like that. We use it in `one-key-formatting`

fregante commented:
> Can't we `onConversationHeader` something like that. We use it in [`one-key-formatting`](/sindresorhus/refined-github/blob/main/source/features/one-key-formatting.tsx)

To be honest I don't really like that file. Here specifically it can be solved by just adding a check for `.suggester`

darkred commented:
> Thank you for the PR and for following the template! I fixed some mistakes but I'm not sure it passes lint/types.

@fregante Always glad to be of help!
Thanks for your commit. I just added a needed type assertion + a missing bracket because it failed lint/types.

yakov116 commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,27 @@
+}
+
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isNewIssue,
+		pageDetect.isCompare,
+		pageDetect.isRepo,
+		pageDetect.isEditingFile
+	],
+	awaitDomReady: false,
```
```suggestion
	include: [
		pageDetect.isNewIssue,
		pageDetect.isPR,
		pageDetect.isEditingFile
	],
```

Does this work?

darkred replied:
No, because `isRepo` is needed for commits when *creating* new files e.g. 
https://github.com/sindresorhus/refined-github/new/main

Checked via https://fregante.github.io/github-url-detection/ : 
![2021-04-19_184538](https://user-images.githubusercontent.com/723651/115265035-82899b80-a13f-11eb-9832-6334f463518b.jpg)

yakov116 replied:
Please open an issue there.

Adding `isRepo` will add it to all repo pages (PR's Issues etc)

darkred replied:
I opened it here: https://github.com/fregante/github-url-detection/issues/68

yakov116 commented:
Blocked by https://github.com/fregante/github-url-detection/issues/68

steel9209 commented:
Yesterday I came across the cloud from garnet about pstn and I had ity and it had to do with a new cloud and new creations with it. And there it is Microsoft edge I think is stealing it and made something out of it. I'm so tired of getting screwed over, and over again. And peoples terms and policies just are unfair. Is there a way to write terms and policies for those who have accessed my data and my files that there be a fee to pay for my service and my technology and royalty or take over ownership of their companies that stole my files and claiming it as theirs, or something...i'm just getting screwed over before I can start. Need to be able to track and trace everything that has gone out to me and stolen from me. Or copies being used, i don't know even the government has my data that they shared to each other and they hold my money from unclaimed assets and more. I don't know what all this means, can I claim to be owner or have ownership in who has my data and using it? There's something rite?

yakov116 reviewed (approved):
Looks great!

darkred commented:
I made a change with <strike>https://github.com/sindresorhus/refined-github/pull/4252/commits/ad2c394aaf253ac9d8a087c63c32f765154bf86e</strike> https://github.com/sindresorhus/refined-github/pull/4252/commits/13762acdf461cd19ac5a9e087ec735523a15b762
in order `isNewFile` and `isEditingFile` to focus the optional description textarea `#commit-description-textarea` :

![](https://user-images.githubusercontent.com/723651/115701515-8c461580-a370-11eb-8612-e0c500c47db6.gif)

fregante commented:
> previously they didn't focus anything

That's impossible. `#commit-description-textarea` is also a `textarea` and the selector is looking in the same form. If `#commit-description-textarea` matches then `textarea` should too

fregante commented:
Ah I see what's going on. In the same form there's another `textarea` for the file’s content, hidden by CodeMirror.

fregante commented on readme.md:
```diff
@@ -176,6 +176,7 @@ Thanks for contributing! 🦋🙌
 - [](# "table-input") [Adds a button in the text editor to quickly insert a simplified HTML table.](https://user-images.githubusercontent.com/46634000/94559114-09892c00-0261-11eb-8fb0-c5a85ea76b6f.gif)
 - [](# "unfinished-comments") [Notifies the user of unfinished comments in hidden tabs.](https://user-images.githubusercontent.com/1402241/97792086-423d5d80-1b9f-11eb-9a3a-daf716d10b0e.gif)
 - [](# "wait-for-attachments") [Wait for the attachments to finish uploading before allowing to post a comment.](https://user-images.githubusercontent.com/46634000/104294547-9b8b0c80-54bf-11eb-93e5-65ae158353b3.gif)
+- [](# "avoid-accidental-submissions") [Re-maps the enter key to act as a tab in issue/PR/commit title fields, so it will merely move the cursor to the textarea, to avoid accidental submissions.](https://user-images.githubusercontent.com/723651/115148453-5818df00-a068-11eb-92d4-51ac6e937b8b.gif)
```
```suggestion
- [](# "avoid-accidental-submissions") [Disables the <kbd>enter</kbd>-to-submit shortcut in some commit/PR/issue title fields to avoid accidental submissions. Use <kbd>ctrl</kbd><kbd>enter</kbd> instead.](https://user-images.githubusercontent.com/723651/115148453-5818df00-a068-11eb-92d4-51ac6e937b8b.gif)
```

darkred replied:
Just one thing to clarify:
by default (=without RG)
in the GitHub pages `isNewFile` and `isEditingFile`, while having focus on `#commit-summary-input`,
submitting via <kbd>ctrl</kbd><kbd>enter</kbd> is not possible: 
one could only submit via <kbd>enter</kbd> or via clicking 'Commit New File'/'Commit changes'

fregante replied:
Sounds like it should be added if it doesn’t work

darkred replied:
I added it in my new commit https://github.com/sindresorhus/refined-github/pull/4252/commits/4ebe35748079e5207afc81c22a6c907205c09137

fregante commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,30 @@
+function onKeyDown(event: delegate.Event<KeyboardEvent, HTMLFormElement>): void {
+	if (event.key === 'Enter') {
+		event.preventDefault();
+		if (!pageDetect.isNewFile() && !pageDetect.isEditingFile()) {
+			select('textarea', event.delegateTarget.form)!.focus();
+		} else {
+			select('#commit-description-textarea', event.delegateTarget.form)!.focus();
+		}
```
```suggestion
		select('[name=description],[name=somethingelse]', event.delegateTarget.form)!.focus();
```

darkred replied:
Fixed with https://github.com/sindresorhus/refined-github/pull/4252/commits/e37d02bd32a8ff37535b215e26b98b93afcc3b78 :
`select('#issue_body, #pull_request_body, #commit-description-textarea', event.delegateTarget.form)!.focus();
`

fregante commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,30 @@
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isNewIssue,
+		() => pageDetect.isCompare() && select.exists('#new_pull_request'),
```
I think
```suggestion
		pageDetect.isQuickPR,
```

darkred replied:
`isQuickPR` ([see](https://github.com/fregante/github-url-detection/blob/master/index.ts#L230)) matches 
`https://github.com/sindresorhus/refined-github/compare/master...branch-name?quick_pull=1`
but not also
`https://github.com/sindresorhus/refined-github/compare/master...branch-name`

fregante replied:
@yakov116 do you remember any tests that detect this? Or is it even needed? Any compare page could show that form I think, if there's any diff

yakov116 replied:
I don't bug I can check.

I think we had a feature that only ran on that page

darkred replied:
To facilitate you, it's only used in this feature: https://github.com/sindresorhus/refined-github/blob/main/source/features/linkify-branch-references.tsx

fregante replied:
I saw something similar in the very similar `prevent-duplicate-pr-submission`


Since it's just an event listener, lets' just use this

```suggestion
		pageDetect.isCompare(),
```

darkred replied:
You meant `pageDetect.isCompare`. Just changed 👍

yakov116 commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,34 @@
+	}
+}
+
+// Add ability to submit via Ctrl+Enter in isNewFile and isEditingFile pages (was not available before)
+function onKeyDownCtrlEnter(event: delegate.Event<KeyboardEvent, HTMLFormElement>): void {
+	if (event.key === 'Enter' && event.ctrlKey) {
+		event.delegateTarget.form.submit();
+	}
+}
```
Why cant we do

https://github.com/sindresorhus/refined-github/blob/8ca75a97bf92ade57b220f9945a06e3eee865c9c/source/features/create-release-shortcut.tsx#L6-L8

darkred replied:
Added in https://github.com/sindresorhus/refined-github/pull/4252/commits/e3645e92814c6aa2874e8871f09faebd7239708c

yakov116 replied:
Can  you add the shortcut (look at that file to see what I did there)

darkred replied:
Added in https://github.com/sindresorhus/refined-github/pull/4252/commits/774a8798c8dc0232b60408e1d96f41992092c6f2

yakov116 commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,39 @@
+	delegate(document, 'input#issue_title, input#pull_request_title, input#commit-summary-input', 'keydown', onKeyDownEnter);
+}
+
+function addQuickSubmit(): void {
```
Move this to on top of onKeyDownEnter

darkred replied:
Done

fregante commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,39 @@
+	init
+}, {
+	shortcuts: {
+		'ctrl enter': 'Publish a new/edited file'
```
Thanks for this, but since it's not actually a RG feature we should not add it here.

yakov116 replied:
Well we added it to the publish release...

fregante commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,39 @@
+import features from '.';
+
+function addQuickSubmit(): void {
+	select('input#commit-summary-input')?.classList.add('js-quick-submit');
```
Why `?.`? When is this missing?

darkred replied:
It missed my attention in the copy-paste. Changed to `!`.

fregante commented:
Thanks @darkred!

darkred commented:
I'm glad I could help!
And, thank you, Federico, and @yakov116 for the guidance 👍

fregante commented on source/features/avoid-accidental-submissions.tsx:
```diff
@@ -0,0 +1,39 @@
+
+function addQuickSubmit(): void {
+	select('input#commit-summary-input')!.classList.add('js-quick-submit');
+}
```
Not sure why we added this here. Doesn't this **allow** submission via ctrl-enter? It should be extracted to its own feature (and the shortcut should mention `cmd` on Mac )
